### PR TITLE
Adding Bright edges contribution

### PR DESCRIPTION
**Pull request by fiammante**
This pull request adds a new way of detecting edges used is low resolution image object recognition in real projects (e.g. 50cm per pixel). It corrects surfaces for objects partially under a lighting shadow, and reveal low visibility edges. The result is used to feed further object detection processes particularly in context where small edges or details are important for object discrimination.
The BrightEdges.cpp provides an implementation of an equalized absolute difference of blurs, and an optional further treatment to contrast edges based on finding local minimum along at least two directions. The local minimum detection contrast is a parameter, using 10 as default (10 on 255 maximum).
The kernel size for the gaussian and the average blur can be passed as parameters too.
brightedgesdemo.cpp provides a sample of different results with or without edge contrasting. Default image used the ml.png .
testbrightedges

```
force_builders=Docs,Linux x64
```

### CODE REVIEW

**Comment by fiammante** [comment]
Corrected hpp file

File: modules/ximgproc/src/brightedges.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+	}
+	
+	int contrastEdges(Mat &minput, Mat &mouput, int contrast) {
+
+		unsigned char *input = (unsigned char*)(mouput.data);
+
+		Mat mwork = minput.clone();
+		mwork.setTo(Scalar(255));
+
+		mouput.setTo(Scalar(255));
```
Comment (terfendail) [inline modules/ximgproc/src/brightedges.cpp]: This assignment could be done later when `mwork` is already filled to allow inplace processing as well.

File: modules/ximgproc/src/brightedges.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+		return input[iedge.cols * (row)+(col)];
+
+
+	}
+	
+	int contrastEdges(Mat &minput, Mat &mouput, int contrast) {
+
+		unsigned char *input = (unsigned char*)(mouput.data);
+
+		Mat mwork = minput.clone();
```
Comment (terfendail) [inline modules/ximgproc/src/brightedges.cpp]: Cloning is excessive here. It would be better to just create the matrix of given size and type.
i.e. `Mat mwork(minput.size(), minput.type(), Scalar(255));`

File: modules/ximgproc/src/brightedges.cpp [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+#include <signal.h>
+#include <opencv2/ximgproc.hpp>
+namespace cv
+{
+	namespace ximgproc
+	{
+	bool isPixelMinimum(Mat &edge, int row, int col, int contrast) {
+		int count = 0;
+		unsigned char *input = (unsigned char*)(edge.data);
+		int pixel = input[edge.cols * row + col] + contrast;
```
Comment (terfendail) [inline modules/ximgproc/src/brightedges.cpp]: It would be better to use `edge.at<unsigned char>(row, col)` or `edge.ptr(row)[col]` to consider the case of non-continuous image(when `edge.step != edge.cols`)

File: modules/ximgproc/src/brightedges.cpp (hunk 1 above)
Comment (fiammante) [reply modules/ximgproc/src/brightedges.cpp]: @terfendail Corrected the code to only explicitly work on the border instead of setting all matrix to 255. Also corrected the minimum detection. (FYI in my project we don't use contrasting as convolutions do the job for us). See the new sample image (I tested on many different images).
testbrightedges

File: modules/ximgproc/src/brightedges.cpp [hunk 4]
```diff
@@ -219,7 +260,7 @@ namespace cv
 
 		absdiff(gblur, bblur, diff);
 
-		cvtColor(diff, gray, COLOR_BGR2GRAY);
+		cvtColor(diff, gray, CV_BGR2GRAY);
```
Comment (sturkmen72) [inline modules/ximgproc/src/brightedges.cpp]: please use COLOR_BGR2GRAY

**Comment by fiammante** [comment]
@sturkmen72 Thank you, fixed.

File: modules/ximgproc/src/brightedges.cpp (hunk 2 above)
Comment (fiammante) [reply modules/ximgproc/src/brightedges.cpp]: Replaced clone thank you.

**Comment by fiammante** [comment]
removed all clones

File: modules/ximgproc/src/brightedges.cpp [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+#include "precomp.hpp"
+#include <iostream>
+#include <signal.h>
+namespace cv
+{
+
+	bool isPixelMinimum(Mat &edge, int row, int col, int contrast) {
+		int count = 0;
+		unsigned char *input = (unsigned char*)(edge.data);
+		int pixel = input[edge.cols * row + col] + contrast - 1;
```
Comment (terfendail) [inline modules/ximgproc/src/brightedges.cpp]: It would be better to use edge.at<unsigned char>(row, col) or edge.ptr(row)[col] to consider the case of non-continuous image(when edge.step != edge.cols)

File: modules/ximgproc/include/opencv2/ximgproc.hpp [hunk 6]
```diff
@@ -56,26 +56,21 @@
...
 
-
 /** @defgroup ximgproc Extended Image Processing
   @{
     @defgroup ximgproc_edge Structured forests for fast edge detection
 
 This module contains implementations of modern structured edge detection algorithms,
 i.e. algorithms which somehow takes into account pixel affinities in natural images.
 
-    @defgroup ximgproc_edgeboxes EdgeBoxes
```
Comment (terfendail) [inline modules/ximgproc/include/opencv2/ximgproc.hpp]: Changes in this file looks unrelated to PR purpose. Are these changes necessary?

**Comment by terfendail** [comment]
@fiammante
Could you please move the change to other branch?
There is an issue with existing CI infrastructure that makes impossible to test changes proposed from `master` branch.

File: modules/ximgproc/include/opencv2/ximgproc.hpp (hunk 6 above)
Comment (fiammante) [reply modules/ximgproc/include/opencv2/ximgproc.hpp]: @tenderfail I added the interface to the brightedge method if any one wants to use it just as the other  do to expose their methods from the src folder. I can leave this hpp alone if you wish.

**Comment by fiammante** [comment]
@terfendail  ok closing this pull request. Which branch should I use?

**Comment by fiammante** [comment]
@terfendail  I only see a master branch here https://github.com/opencv/opencv_contrib how can I use another branch?

File: modules/ximgproc/src/brightedges.cpp (hunk 5 above)
Comment (fiammante) [reply modules/ximgproc/src/brightedges.cpp]: @terfendail Sounds strange to me. I used to develop device driver and pointer arithmetic (array indexing) was always much compact faster looking at the assembler generated by C++ (I debugged assembler from device drivers written in C++). Is there a particular code with edge.at() that makes it safer?

**Comment by alalek** [comment]
It is about the source branch of PR, not the target branch which is always master for opencv_contrib.
```
	fiammante  wants to merge 7 commits into opencv:master from fiammante:master
                                                                               ^^^
                                                                             fix this
```

Unfortunately GitHub doesn't allow to do this without opening a new pull request from proper branch.

**Comment by fiammante** [comment]
@alalek can you please tell me which source branch I should use?

File: modules/ximgproc/src/brightedges.cpp (hunk 5 above)
Comment (terfendail) [reply modules/ximgproc/src/brightedges.cpp]: OpenCV allows matrix data to be non-continuous so the next matrix row isn't always located just after the previous in memory(it could be due to alignment reasons or specific preprocessing of source matrix). There is a stride field in the matrix header that specify step between rows and edge.at() consider the matrix specific when performs address arithmetic.

File: modules/ximgproc/include/opencv2/ximgproc.hpp (hunk 6 above)
Comment (terfendail) [reply modules/ximgproc/include/opencv2/ximgproc.hpp]: Lines changed in the file are related to the documentation structure. Also removed documentation groups are used in descriptions of other algorithms so the change could even affect documentation build. Thus IMHO these changes should be reverted.

**Comment by terfendail** [comment]
@fiammante you could use any branch name other than master as a source PR branch.
i.e you could create `bright_edges` branch in your repository that will point to the same commit as your master branch

File: modules/ximgproc/src/brightedges.cpp (hunk 5 above)
Comment (fiammante) [reply modules/ximgproc/src/brightedges.cpp]: @tenderfail thank you understood. I shall change code accordingly, but then which branch do I put it in? In the ximgproc samples folder? Thank you in advance
